src/map/mod.rs

```diff
@@ -0,0 +1,201 @@
+use std::marker::PhantomData;
+
+use wasm_bindgen_futures::spawn_local;
+use web_sys::wasm_bindgen::JsCast;
+use xilem_web::{
+    core::{MessageResult, Mut, View, ViewId, ViewMarker},
+    DomNode, DynMessage, ElementProps, Pod, ViewCtx, WithStyle, HTML_NS,
+};
+
+mod events;
+pub use self::events::*;
+
+// TODO:
+// How can we pass tuples, like `()`, `(C)`, `(C, C)`?
+pub fn map<State, Action, C>(children: Vec<C>) -> Map<State, Action, C>
+where
+    C: MapChild<State, Action>,
+    C::ViewState: MapChildViewState,
+{
+    Map {
+        zoom: None,
+        center: None,
+        children,
+        phantom: PhantomData,
+    }
+}
+
+/// This is a marker trait that is used to pass only map-specific children
+/// to the [`map`] function.
+pub trait MapChild<State, Action>: View<State, Action, ViewCtx, DynMessage>
+where
+    Self::ViewState: MapChildViewState,
+{
+}
+
+/// This is a hack that allows children of the map
+/// to get access to [`leaflet::Map`].
+pub trait MapChildViewState {
+    fn after_build(&mut self, map: &leaflet::Map);
+    fn after_rebuild(&mut self, map: &leaflet::Map);
+}
+
+pub struct Map<State, Action, C>
+where
+    C: MapChild<State, Action>,
+    C::ViewState: MapChildViewState,
+{
+    children: Vec<C>,
+    zoom: Option<f64>,
+    center: Option<(f64, f64)>,
+    phantom: PhantomData<fn() -> (State, Action)>,
+}
+
+impl<State, Action, C> Map<State, Action, C>
+where
+    C: MapChild<State, Action>,
+    C::ViewState: MapChildViewState,
+{
+    pub fn zoom(mut self, value: f64) -> Self {
+        self.zoom = Some(value);
+        self
+    }
+
+    pub fn on_zoom_end<F>(&self, callback: F) -> OnZoomEnd<State, Action, F>
+    where
+        F: Fn(&mut State, f64) + 'static,
+    {
+        on_zoom_end(callback)
+    }
+
+    pub fn center(mut self, lat: f64, lng: f64) -> Self {
+        self.center = Some((lat, lng));
+        self
+    }
+}
+
+impl<State, Action, C> ViewMarker for Map<State, Action, C>
+where
+    C: MapChild<State, Action>,
+    C::ViewState: MapChildViewState,
+{
+}
+
+pub struct MapViewState<State, Action, C>
+where
+    C: MapChild<State, Action>,
+    C::ViewState: MapChildViewState,
+{
+    leaflet_map: leaflet::Map,
+    children: Vec<(C::Element, C::ViewState)>,
+}
+
+#[derive(Debug)]
+enum MapMessage {
+    InitMap,
+}
+
+impl<State, Action, C> View<State, Action, ViewCtx, DynMessage> for Map<State, Action, C>
+where
+    State: 'static,
+    Action: 'static,
+    C: MapChild<State, Action>,
+    C::ViewState: MapChildViewState,
+{
+    type Element = Pod<web_sys::HtmlElement, ElementProps>;
+
+    type ViewState = MapViewState<State, Action, C>;
+
+    fn build(&self, ctx: &mut ViewCtx) -> (Self::Element, Self::ViewState) {
+        // TODO:
+        // Is there a way to just use `html::div(())`?
```

Comment:
Yes there is, it's a little bit annoying without TAITs to do this, as the concrete type needs to be defined. (See also #2, for the `Frozen` view which *should* be used there TAITs would be more or less necessary)

There was also [this suggestion](https://github.com/DJMcNab/xilem/pull/4), and I hope we can get to that direction in one way or the other (but that had issues, because of an `OrphanImpl` Marker in `View`, not being compatible with traits derived of `View`).